File: packages/core/src/extract.ts
/**
 * SQL extraction from model output.
 * Robust to fenced blocks, prose wrapping, and multiple fences.
 */

// Any language tag is consumed; the candidate is still gated by SQL_START_RE.
const FENCE_RE = /```[A-Za-z0-9+_-]*\s*\n?([\s\S]*?)```/gu;

/** Any statement-shaped start, including write/DDL verbs: the guard, not the extractor, decides what may run. */
const SQL_START_RE =
  /^(select|with|explain|show|describe|desc|pragma|insert|update|delete|drop|create|alter|truncate|merge|replace|call|grant|revoke|copy|values|table)\b/iu;

/** Conservative set for inline extraction from prose - read verbs only. */
const INLINE_START_RE = /(?:^|\n)\s*((?:select|with|explain)\b[\s\S]*?)(?=\n\s*\n|$)/iu;

export interface Extraction {
  readonly sql: string;
  readonly explanation: string;
  readonly source: 'fence' | 'inline' | 'whole';
}

/** Sentinel the prompt asks the model to emit when a question is unanswerable. */
export const IMPOSSIBLE_SENTINEL = /^\s*IMPOSSIBLE\s*:\s*(.+)/su;

const REASON_MAX_LENGTH = 300;
/** The sentinel word is internal protocol; a model that repeats it mid-sentence must not leak it into the chat. */
const SENTINEL_WORD = /\bIMPOSSIBLE\b\s*:?\s*/gi;
/** "Your question isn't about this data" said many robotic ways; all collapse to one plain sentence. */
const OFF_TOPIC =
  /\b(the )?question (cannot be answered|is not|isn't)\b[^.]*\b(not related to|unrelated to|does not relate)\b|\bnot related to the (provided )?schema\b/i;
/** Model-speak to plain English, applied in order. Deterministic, no second model call. */
const PHRASINGS: readonly (readonly [RegExp, string])[] = [
  [/\bthe provided schema\b/gi, 'this database'],
  [/\bthe (given |current )?schema\b/gi, 'this database'],
  [/\bdoes not contain any information (about|on|related to)\b/gi, "doesn't have anything about"],
  [/\bdoes not contain any\b/gi, "doesn't have any"],
  [/\bdoes not contain\b/gi, "doesn't have"],
  [/\bdoes not (include|have|provide)\b/gi, "doesn't have"],
  [/\bis not able to\b|\bcannot be\b/gi, "can't be"],
];

function humanizeReason(reason: string): string {
  if (OFF_TOPIC.test(reason)) return "That question isn't about the data in this database.";
  let out = reason;
  for (const [pattern, replacement] of PHRASINGS) out = out.replace(pattern, replacement);
  return out.replace(/\s{2,}/g, ' ').trim();
}

function truncateAtWordBoundary(text: string, max: number): string {
  if (text.length <= max) return text;
  const cut = text.slice(0, max);
  const lastSpace = cut.lastIndexOf(' ');
  return `${(lastSpace > max / 2 ? cut.slice(0, lastSpace) : cut).trimEnd()}…`;
}

/** The first line after "IMPOSSIBLE:" is the reason; the sentinel is stripped and stiff phrasing humanized. */
export function extractImpossible(text: string): string | null {
  const m = IMPOSSIBLE_SENTINEL.exec(withoutReasoning(text).trim());
  if (!m) return null;
  const firstLine = m[1]!.trim().split('\n')[0]!.trim();
  const cleaned = firstLine.replace(SENTINEL_WORD, '').trim();
  const humanized = humanizeReason(cleaned);
  const sentenceCased = humanized.charAt(0).toUpperCase() + humanized.slice(1);
  return truncateAtWordBoundary(sentenceCased, REASON_MAX_LENGTH);
}

/**
 * A reasoning model narrates before it answers. Groq's qwen3.6 opens with "<think>\nHere's a thinking
 * process:", and that text was shown to the reader as the query's description. Worse, an answer cut off
 * mid-reasoning leaves the tag unclosed, so everything after it is narration with no answer in it.
 */
const THINK_BLOCK = /<(think|thinking|reasoning)>[\s\S]*?<\/\1>/giu;
// Anchored: unanchored, a tag inside the answer truncated `WHERE body LIKE '%<think>%'` to an
// unterminated literal.
const THINK_UNCLOSED = /^\s*<(?:think|thinking|reasoning)>[\s\S]*$/iu;

/**
 * Hides a reasoning model's narration as it streams; the whole-text strip only cleans the assembled
 * reply, so a streaming host forwarded the monologue live. Only a tag that OPENS the reply counts: one
 * appearing later is content. A tag split across chunks is held back until it can be read.
 */
export function createReasoningFilter(): (chunk: string) => string {
  const OPEN = /<(think|thinking|reasoning)>/i;
  const CLOSE = /<\/(think|thinking|reasoning)>/i;
  const LONGEST_TAG = '</reasoning>'.length;
  let phase: 'leading' | 'narrating' | 'passthrough' = 'leading';
  let carry = '';

  return (chunk: string): string => {
    let buffer = carry + chunk;
    carry = '';

    if (phase === 'leading') {
      const open = OPEN.exec(buffer);
      if (open && buffer.slice(0, open.index).trim() === '') {
        buffer = buffer.slice(open.index + open[0].length);
        phase = 'narrating';
      } else if (buffer.trim() === '' || (!open && buffer.trimStart().startsWith('<') && buffer.length < LONGEST_TAG)) {
        carry = buffer; // could still become an opening tag
        return '';
      } else {
        phase = 'passthrough';
      }
    }

    if (phase === 'narrating') {
      const close = CLOSE.exec(buffer);
      if (!close) {
        const cut = buffer.lastIndexOf('<');
        carry = cut >= 0 && buffer.length - cut <= LONGEST_TAG ? buffer.slice(cut) : '';
        return '';
      }
      buffer = buffer.slice(close.index + close[0].length);
      phase = 'passthrough';
    }

    return buffer;
  };
}

/** Removes a reasoning model's narration, leaving the answer it was working towards. */
export function withoutReasoning(text: string): string {
  return text.replace(THINK_BLOCK, ' ').replace(THINK_UNCLOSED, ' ').trim();
}

export function extractSql(text: string): Extraction | null {
  const raw = withoutReasoning(text ?? '');

  // 1) Fenced blocks - first block that looks like a query wins.
  const fences = [...raw.matchAll(FENCE_RE)];
  for (const f of fences) {
    const candidate = (f[1] ?? '').trim();
    if (candidate && SQL_START_RE.test(candidate)) {
      const explanation = raw.replace(f[0]!, ' ').replace(/```[\s\S]*?```/gu, ' ');
      return { sql: candidate, explanation: tidy(explanation), source: 'fence' };
    }
  }

  // 2) Whole message is SQL.
  const trimmed = raw.trim();
  if (SQL_START_RE.test(trimmed)) {
    return { sql: trimmed, explanation: '', source: 'whole' };
  }

  // 3) Inline: first SELECT/WITH run up to a blank line or end.
  const inline = INLINE_START_RE.exec(raw);
  if (inline) {
    const sql = inline[1]!.trim();
    if (sql.length > 8) {
      return { sql, explanation: tidy(raw.replace(inline[1]!, ' ')), source: 'inline' };
    }
  }
  return null;
}

/** A hedging "IMPOSSIBLE:" line. The query wins (see extractSql), so the hedge must not travel with it. */
const SENTINEL_LINE = /^[^\S\n]*IMPOSSIBLE\b[^\n]*$/gimu;

// Normalize whitespace; no length cap - the description stays complete.
function tidy(explanation: string): string {
  return explanation.replace(SENTINEL_LINE, ' ').replace(SENTINEL_WORD, '').replace(/\s+/gu, ' ').trim();
}

File: packages/jetbrains/src/main/kotlin/com/rahulmahadik/asksql/ide/engine/Extract.kt
package com.rahulmahadik.asksql.ide.engine

/** SQL extraction from model output (same behavior as core's `extract.ts`): fenced blocks, prose wrapping, multiple fences. */
object Extract {

    enum class ExtractionSource { FENCE, INLINE, WHOLE }

    data class Extraction(val sql: String, val explanation: String, val source: ExtractionSource = ExtractionSource.FENCE)

    // \w* (not a literal "sql"/"SQL") matches ANY fence language tag, so
    // ```postgresql, ```MySQL, ```sqlite, etc. are still recognized as fenced blocks.
    private val FENCE_RE = Regex("""```\w*\s*\n?([\s\S]*?)```""")

    /** Any statement-shaped start, including write/DDL verbs: the guard, not the extractor, decides what may run. */
    private val SQL_START_RE = Regex(
        """^(select|with|explain|show|describe|desc|pragma|insert|update|delete|drop|create|alter|truncate|merge|replace|call|grant|revoke|copy|values|table)\b""",
        RegexOption.IGNORE_CASE,
    )

    /** Conservative set for INLINE extraction from prose: read verbs only, to avoid grabbing English sentences that begin with "Update"/"Insert". */
    private val INLINE_START_RE = Regex(
        """(?:^|\n)\s*((?:select|with|explain)\b[\s\S]*?)(?=\n\s*\n|$)""",
        RegexOption.IGNORE_CASE,
    )

    // Case-sensitive to match core exactly: the model emits this sentinel verbatim in
    // uppercase, so a lowercase "impossible:" is ordinary prose, not the sentinel.
    private val IMPOSSIBLE_SENTINEL = Regex(
        """^\s*IMPOSSIBLE\s*:\s*(.+)""",
        RegexOption.DOT_MATCHES_ALL,
    )

    /** Trims a fence's trailing `-- Explanation:`/`-- Note:` commentary, for a model that writes prose without closing the fence first. */
    private val TRAILING_PROSE_RE = Regex("""\n\s*--\s*(Explanation|Note)\b.*""", setOf(RegexOption.IGNORE_CASE, RegexOption.DOT_MATCHES_ALL))

    private val REFUSAL = Regex(
        """\b(i can(?:no|')t|i cannot|i am unable|i'm unable|i'm sorry|as an ai)\b""",
        RegexOption.IGNORE_CASE,
    )

    private const val REASON_MAX_LENGTH = 300

    /** The sentinel word is internal protocol; a model that repeats it mid-sentence must not leak it into the chat. */
    private val SENTINEL_WORD = Regex("""\bIMPOSSIBLE\b\s*:?\s*""", RegexOption.IGNORE_CASE)

    /** "Your question isn't about this data" said many robotic ways; all of them collapse to one plain sentence. */
    private val OFF_TOPIC = Regex(
        """\b(the )?question (cannot be answered|is not|isn't)\b[^.]*\b(not related to|unrelated to|does not relate)\b|""" +
            """\bnot related to the (provided )?schema\b""",
        RegexOption.IGNORE_CASE,
    )

    /** Model-speak to plain English, applied in order. Deterministic and local: no second model call. */
    private val PHRASINGS = listOf(
        Regex("""\bthe provided schema\b""", RegexOption.IGNORE_CASE) to "this database",
        Regex("""\bthe (given |current )?schema\b""", RegexOption.IGNORE_CASE) to "this database",
        Regex("""\bdoes not contain any information (about|on|related to)\b""", RegexOption.IGNORE_CASE) to "doesn't have anything about",
        Regex("""\bdoes not contain any\b""", RegexOption.IGNORE_CASE) to "doesn't have any",
        Regex("""\bdoes not contain\b""", RegexOption.IGNORE_CASE) to "doesn't have",
        Regex("""\bdoes not (include|have|provide)\b""", RegexOption.IGNORE_CASE) to "doesn't have",
        Regex("""\bis not able to\b|\bcannot be\b""", RegexOption.IGNORE_CASE) to "can't be",
    )

    /** Rewrites a model's stiff refusal into something readable, keeping its specifics. */
    private fun humanizeReason(reason: String): String {
        if (OFF_TOPIC.containsMatchIn(reason)) return "That question isn't about the data in this database."
        var out = reason
        for ((pattern, replacement) in PHRASINGS) out = pattern.replace(out, replacement)
        return out.replace(Regex("""\s{2,}"""), " ").trim()
    }

    /** The prompt asks for "IMPOSSIBLE: <one-line reason>"; a noncompliant model rambles on, so only the first line is the reason. */
    fun extractImpossible(text: String): String? {
        val captured = IMPOSSIBLE_SENTINEL.find(text.trim())?.groupValues?.get(1)?.trim() ?: return null
        val firstLine = captured.substringBefore('\n').trim()
        val cleaned = SENTINEL_WORD.replace(firstLine, "").trim()
        val humanized = humanizeReason(cleaned).replaceFirstChar { it.uppercase() }
        return truncateAtWordBoundary(humanized, REASON_MAX_LENGTH)
    }

    private fun truncateAtWordBoundary(text: String, maxLength: Int): String {
        if (text.length <= maxLength) return text
        val cut = text.take(maxLength)
        val lastSpace = cut.lastIndexOf(' ')
        return (if (lastSpace > maxLength / 2) cut.take(lastSpace) else cut).trimEnd() + "…"
    }

    fun extractSql(text: String): Extraction? {
        // 1) Fenced blocks: first block that looks like a query wins.
        for (f in FENCE_RE.findAll(text)) {
            var candidate = f.groupValues[1].trim()
            if (candidate.isNotEmpty() && SQL_START_RE.containsMatchIn(candidate)) {
                val trailingProse = TRAILING_PROSE_RE.find(candidate)
                if (trailingProse != null) candidate = candidate.substring(0, trailingProse.range.first).trimEnd()
                val explanation = text.replaceFirst(f.value, " ").replace(Regex("""```[\s\S]*?```"""), " ")
                return Extraction(candidate, tidy(explanation), ExtractionSource.FENCE)
            }
        }

        // 2) Whole message is SQL.
        val trimmed = text.trim()
        if (SQL_START_RE.containsMatchIn(trimmed)) {
            return Extraction(trimmed, "", ExtractionSource.WHOLE)
        }

        // 3) Inline: first SELECT/WITH/EXPLAIN run up to a blank line or end.
        val inline = INLINE_START_RE.find(text)
        if (inline != null) {
            val sql = inline.groupValues[1].trim()
            if (sql.length > 8) {
                return Extraction(sql, tidy(text.replaceFirst(inline.groupValues[1], " ")), ExtractionSource.INLINE)
            }
        }
        return null
    }

    fun looksLikeRefusal(text: String): Boolean = REFUSAL.containsMatchIn(text)

    private fun tidy(explanation: String): String =
        explanation.replace(Regex("""\s+"""), " ").trim().take(2000)
}
